### PR TITLE
TST: Missing polynomial tests

### PR DESCRIPTION
**Pull request by jeffyancey**
- Added tests for chebypow, hermepow, hermpow, lagpow, legpow, and polypow. 
- Added missing methods to various `See Also` sections
- Added missing methods to arithmetic in docstring.
- Added `Examples` where missing

### CODE REVIEW

File: numpy/polynomial/tests/test_polynomial.py [hunk 1]
```diff
@@ -102,6 +103,19 @@ def test_polydiv(self):
                 res = poly.polyadd(poly.polymul(quo, ci), rem)
                 assert_equal(res, tgt, err_msg=msg)
 
+    def test_polypow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = list(range(i+1))
+                [c] = pu.as_series([c])
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_polynomial.py]: Why do you need `as_series` here? None of the other tests do it.

File: numpy/polynomial/polynomial.py [hunk 2]
```diff
@@ -450,10 +455,13 @@ def polypow(c, pow, maxpower=None):
 
     See Also
     --------
-    polyadd, polysub, polymul, polydiv
+    polyadd, polysub,polymulx, polymul, polydiv
```
Comment (eric-wieser) [inline numpy/polynomial/polynomial.py]: nit: missing space after comma

**Review by eric-wieser — COMMENTED** [review]
A few remarks, but generally looks good.

A note about commit messages: Can you change them to use the standard prefixes? I'd imagine that you can categorize these as a mix of `DOC` and `TST`

File: numpy/polynomial/tests/test_polynomial.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/polynomial/tests/test_polynomial.py]: Perhaps just `c = np.arange(1, i + 1)` in place of both these lines?

File: numpy/polynomial/tests/test_chebyshev.py [hunk 3]
```diff
@@ -111,6 +112,21 @@ def test_chebdiv(self):
...
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = list(range(i+1))
+                [c] = pu.as_series([c])
+                power = j+1
+                zs = cheb._cseries_to_zseries(c)
+                prd = zs
+                for i in range(2, power + 1):
+                    prd = np.convolve(prd, zs)
```
[the remark this answers is not in the document]
Comment (jeffyancey) [reply numpy/polynomial/tests/test_chebyshev.py]: Hi @eric-wieser, thanks for the review. I worked my way down alphabetically, so I began with this function, but Hermite_E did not have the `_cseries_to_zseries` method, making reuse of this test unfeasible. So, I decided it would probably be best to stick as closely as possible to how each `pow` function performed its calculation.

File: numpy/polynomial/tests/test_polynomial.py (hunk 1 above)
Comment (jeffyancey) [reply numpy/polynomial/tests/test_polynomial.py]: That sounds good to me. Thanks for the recommendation.

**Comment by eric-wieser** [comment]
The PR title doesn't matter so much, as we can edit that anyway - its the title of the commits themselves that I'm referring to

The easy option is for me to just squash all the commits together when I merge them through the web UI - in this case, the changes are similar enough that it's fine I think to lose the information the separate commits provide.

The harder option, if you want to practice a little with git on a simple history, is for you to use `git rebase -i` to reword and/or combine the past commits.

File: numpy/polynomial/tests/test_chebyshev.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/polynomial/tests/test_chebyshev.py]: Ideally your tests shouldn't match your implementation - writing the same function in two places doesn't really test anything - and you probably shouldn't be calling private (underscore-led) functions in these tests.

As someone reading the tests, I can look at `test_hermpow` and think _"yes, I expect power to be repeated multiplication"_. I need to know about internal implementation details to understand what this test does.

[the remark this answers is not in the document]
Comment (jeffyancey) [reply numpy/polynomial/tests/test_chebyshev.py]: Makes sense to me, and I should've realize using private methods doesn't really make sense. I'll give this a little ponder and see if I can't come up with something more appropriate.

File: numpy/polynomial/tests/test_laguerre.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 import numpy.polynomial.laguerre as lag
+import numpy.polynomial.polyutils as pu
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_laguerre.py]: This is now unused

Comment (jeffyancey) [reply numpy/polynomial/tests/test_laguerre.py]: good eye!

File: numpy/polynomial/tests/test_polynomial.py [hunk 5]
```diff
@@ -102,6 +102,18 @@ def test_polydiv(self):
                 res = poly.polyadd(poly.polymul(quo, ci), rem)
                 assert_equal(res, tgt, err_msg=msg)
 
+    def test_polypow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.random.randint(0, 100, i + 1)
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_polynomial.py]: I worry that on 32-bit platforms this could overflow - `100 ** 5` doesn't fit in a 32-bit integer.

Are you sure you need a random test here? None of the other tests are random

Comment (jeffyancey) [reply numpy/polynomial/tests/test_polynomial.py]: I don't suppose they need to be, no. I saw some of the other tests in the package were and thought it might be a good idea, thanks for pointing out why it isn't.

File: numpy/polynomial/tests/test_legendre.py [hunk 6]
```diff
@@ -100,6 +100,18 @@ def test_legdiv(self):
...
                 assert_equal(trim(res), trim(tgt), err_msg=msg)
 
+    def test_legpow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.array(range(i + 1))
+                power = j + 1
+                tgt = c
+                for i in range(2, power + 1):
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_legendre.py]: This overwrites the `i` variable, which could end up confusing someone editing this test. Why not `range(1, power)` also?

File: numpy/polynomial/tests/test_chebyshev.py [hunk 7]
```diff
@@ -111,6 +113,16 @@ def test_chebdiv(self):
                 res = cheb.chebadd(cheb.chebmul(quo, ci), rem)
                 assert_equal(trim(res), trim(tgt), err_msg=msg)
 
+    def test_chebpow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.array(range(i + 1))
+                power = j + 1
+                tgt = reduce(cheb.chebmul, [c]*power)
```
Comment (jeffyancey) [inline numpy/polynomial/tests/test_chebyshev.py]: @eric-wieser, after your comment, I realized i could replace `i` with `_`, since it was not being used, then I wondered why we weren't using reduce, and if it would take a performance hit, so I tried it, and as best as I can tell from lot of `timeit` testing, they are roughly equivlent, though the `reduce` seemed to be performing better as the power grew. Please let me know if we are discouraged from using functools/itertools/collections modules, and I will go back to the for loop and do as you recommended.

Comment (eric-wieser) [reply numpy/polynomial/tests/test_chebyshev.py]: Using `reduce` here is fine by me.

File: numpy/polynomial/tests/test_hermite.py [hunk 8]
```diff
@@ -99,6 +101,16 @@ def test_hermdiv(self):
                 res = herm.hermadd(herm.hermmul(quo, ci), rem)
                 assert_equal(trim(res), trim(tgt), err_msg=msg)
 
+    def test_hermpow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.array(range(i + 1))
+                power = j + 1
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_hermite.py]: Sorry to keep spotting things - but can you remove the `+1` in both places here? We should be testing that `pow(c, 0)` works too (for all of these cases).

Comment (jeffyancey) [reply numpy/polynomial/tests/test_hermite.py]: Thanks for catching all these things, I appreciate it.

**Comment by ewmoore** [comment]
FWIW np.array(range()) should generally be spelled np.arange.

File: numpy/polynomial/tests/test_hermite.py [hunk 9]
```diff
@@ -99,6 +101,15 @@ def test_hermdiv(self):
                 res = herm.hermadd(herm.hermmul(quo, ci), rem)
                 assert_equal(trim(res), trim(tgt), err_msg=msg)
 
+    def test_hermpow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.arange(i + 1)
+                tgt = reduce(herm.hermmul, [c]*j) if j else np.array([1])
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_hermite.py]: You can spell this `reduce(herm.hermmul, [c]*j, np.array([1]))`

File: numpy/polynomial/tests/test_hermite.py [hunk 10]
```diff
@@ -99,6 +101,15 @@ def test_hermdiv(self):
                 res = herm.hermadd(herm.hermmul(quo, ci), rem)
                 assert_equal(trim(res), trim(tgt), err_msg=msg)
 
+    def test_hermpow(self):
+        for i in range(5):
+            for j in range(5):
+                msg = "At i=%d, j=%d" % (i, j)
+                c = np.arange(i + 1)
```
Comment (eric-wieser) [inline numpy/polynomial/tests/test_hermite.py]: Does this still pass if you use `arange(i)`?

File: numpy/polynomial/polynomial.py [hunk 11]
```diff
@@ -450,10 +455,13 @@ def polypow(c, pow, maxpower=None):
 
     See Also
     --------
-    polyadd, polysub, polymul, polydiv
+    polyadd, polysub, polymulx, polymul, polydiv
 
     Examples
     --------
+    >>> from numpy.polynomial import polynomial as P
+    >>> P.polypow([1,2,3])
```
Comment (eric-wieser) [inline numpy/polynomial/polynomial.py]: Isn't this missing a second argument?

**Comment by eric-wieser** [comment]
This would be easier to review if you put all the doc fixes in a separate CR to the tests

File: numpy/polynomial/polynomial.py (hunk 11 above)
Comment (jeffyancey) [reply numpy/polynomial/polynomial.py]: Yup

File: numpy/polynomial/tests/test_hermite.py (hunk 10 above)
Comment (jeffyancey) [reply numpy/polynomial/tests/test_hermite.py]: No, I'll paste the message if you like

**Comment by jeffyancey** [comment]
Happy to split this into two PRs

**Comment by jeffyancey** [comment]
Closing this PR and opening two new PRs, one for the tests and one for the documentation changes.
